### ISSUE TITLE
JIT: strange timing with logical operators

### ISSUE THREAD

**Issue by wpalfi** [pytorch/pytorch#70572]
### 🐛 Describe the bug

`ScriptFunction`s are very slow in this example. (At least the first invocation for a given input shape). The problem disappears when I use `torch.logical_and()` or `torch.bitwise_and()` instead of the `&`-operator. `jit.script()` gives similar results.

```python
import torch
from time import time

def fun(a):
    b = a==0
    return b & b

a = torch.zeros(3,device='cuda')
fun_trace = torch.jit.trace(fun,a)

for _ in range(4):
    for n in range(1,6):
        torch.cuda.synchronize()
        t = time()
        fun_trace(torch.zeros(n,device='cuda'))
        torch.cuda.synchronize()
        print(f'{n=} {int((time()-t)*1e3)} ms')
```

Output:

```
n=1 103 ms   !
n=2 89 ms   !
n=3 0 ms    <-- example_input for jit.trace()
n=4 91 ms   !
n=5 87 ms   !
n=1 0 ms
n=2 0 ms
n=3 0 ms
n=4 0 ms
n=5 90 ms   ??
n=1 0 ms
n=2 0 ms
n=3 0 ms
n=4 0 ms
n=5 0 ms
n=1 0 ms
n=2 0 ms
n=3 0 ms
n=4 0 ms
n=5 0 ms
```

### Versions

PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.21.4
Libc version: N/A

Python version: 3.9.7 (default, Sep 16 2021, 16:59:28) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22000-SP0
[pip3] torchvision==0.11.1
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.3.1               h59b6b97_2
[conda] mkl                       2021.4.0           haa95532_640
[conda] mkl-service               2.4.0            py39h2bbff1b_0
[conda] mkl_fft                   1.3.1            py39h277e83a_0
[conda] mkl_random                1.2.2            py39hf11a4ad_0
[conda] numpy                     1.21.2           py39hfca59bb_0
[conda] numpy-base                1.21.2           py39h0829f74_0
[conda] pytorch                   1.10.0          py3.9_cuda11.3_cudnn8_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] pytorch-scatter           2.0.9           py39_torch_1.10.0_cu113    pyg
[conda] pytorch-sparse            0.6.12          py39_torch_1.10.0_cu113    pyg
[conda] torchvision               0.11.1               py39_cu113    pytorch

**Comment by davidberard98** [comment]
@wpalfi, thanks for the report.
 
This is a known issue. In script/trace mode, pytorch will do some warmup/profiling runs to improve future performance. In the case of eq + bitwise-and, it is able to fuse the operators together and compile a new function that runs both ops together.  You can read about it here https://dev-discuss.pytorch.org/t/nnc-walkthrough-how-pytorch-ops-get-fused/125. In the case of logical_and, it seems like this fusion doesn’t occur so it doesn’t have the behavior you observed above. The long runtimes on the first few runs are caused by compilation of the new fused op.
 
In this instance, the fused op can only work on a fixed size tensor. So in this case (i.e. where a size-3 tensor is used for tracing), the initial specialization occurs for tensors of size 3. Then when the size 1 tensor runs, a new operator gets compiled; when a size-2 tensor runs a new operator gets compiled, etc. In your example, once all of the different sizes (1 through 5) are attempted, then an operator for each of the sizes exists and the remaining operations run quickly. Recompilation will happen up to 20 times (i.e. for 20 unique tensor shapes), after which the recompilations will stop and the un-optimized graph will run instead.
 
If you're concerned about the performance impact from these initial warmup runs, here are some potential solutions:
* You can leave it as is. After 20 recompilations, it will stop recompiling when it encounters new sizes. You can change the number of recompilations with `torch._C._jit_set_bailout_depth()`
* You can disable nnc entirely with `torch._C._jit_set_texpr_fuser_enabled(False)`. Unfortunately it would disable all nnc fusions, including other fusions that might be improving performance
* You can increase the number of profiling runs with `torch._C._jit_set_num_profiled_runs(2)`. For context: during profiling runs, information about tensor shapes is gathered and later used to compile the fused op. By default, only one profiling run is used (since tensor shapes often stay the same in between executions of the scripted function). In this case, if you increase the number of profiling runs to 2, then the recorded shape information will show that the shape of the input tensor changes; and this would prevent NNC from compiling an over-specialized fused op.

There are also some future changes that could fix issues like this, including NVFuser (a new gpu fuser) and dynamic shapes. See #52286 for some comments on that.

cc @Krovatkin

**Comment by wpalfi** [comment]
@davidberard98 thanks, now its clear. Would be great to have some hints about all of that in the docs. I will try dynamic shapes, exciting news:-).
